server/routerlicious/packages/lambdas-driver/src/document-router/contextManager.ts

```diff
@@ -78,24 +92,47 @@ export class DocumentContextManager extends EventEmitter {
 	}
 
 	/**
-	 * Updates the head to the new offset. The head offset will not be updated if it stays the same or moves backwards.
+	 * Updates the head to the new offset. The head offset will not be updated if it stays the same or moves backwards, except if the allowBackToOffset is specified.
+	 * allowBackToOffset is specified during resume after a lambda pause (eg: circuit breaker)
 	 * @returns True if the head was updated, false if it was not.
 	 */
-	public setHead(head: IQueuedMessage) {
-		if (head.offset > this.head.offset) {
+	public setHead(head: IQueuedMessage, allowBackToOffset?: number | undefined) {
+		// if allowBackToOffset is set and is lower than lastCheckpoint, then dont reprocess and return early
+		if (allowBackToOffset !== undefined && allowBackToOffset <= this.lastCheckpoint.offset) {
+			Lumberjack.info("Not updating contextManager head and returning early", {
+				allowBackToOffset,
+				lastCheckpointOffset: this.lastCheckpoint.offset,
+			});
+			return false;
+		}
+		if (head.offset > this.head.offset || head.offset === allowBackToOffset) {
+			if (head.offset <= this.head.offset) {
+				Lumberjack.info(
+					"Allowing the contextManager head to move to the specified offset",
+					{ allowBackToOffset, currentHeadOffset: this.head.offset },
+				);
+			}
 			this.head = head;
 			return true;
 		}
 
 		return false;
 	}
 
-	public setTail(tail: IQueuedMessage) {
+	public setTail(tail: IQueuedMessage, allowBackToOffset?: number | undefined) {
 		assert(
-			tail.offset > this.tail.offset && tail.offset <= this.head.offset,
-			`${tail.offset} > ${this.tail.offset} && ${tail.offset} <= ${this.head.offset}`,
+			(tail.offset > this.tail.offset || tail.offset === allowBackToOffset) &&
+				tail.offset <= this.head.offset,
+			`(${tail.offset} > ${this.tail.offset} || ${tail.offset} === ${allowBackToOffset}) && ${tail.offset} <= ${this.head.offset}`,
```

Comment:
I agree with the premise of this suggestion. It would be better, IMO, to have the assertion message be a human-readable description of what is expected, rather than a copy-paste of the equation that failed.

Reply:
I agree, this was to follow the format of all the asserts we have in contextManager and documentContext. But I will update these.